python/streamlit/src/ChemiscopeComponent.ts

```diff
@@ -0,0 +1,426 @@
+import { RenderData, Streamlit } from 'streamlit-component-lib';
+import {
+    CONFIG,
+    applyHeightPolicy,
+    applyWidthPolicy,
+    displayWarning,
+    generateHTMLForMode,
+    getOrCreateRoot,
+    toggleLoadingVisible,
+} from './dom-utils';
+import {
+    Dataset,
+    DefaultVisualizer,
+    EnvironmentIndexer,
+    GUID,
+    Indexes,
+    MapVisualizer,
+    Settings,
+    StructureVisualizer,
+    Warnings,
+} from 'chemiscope';
+
+type ChemiscopeMode = 'default' | 'structure' | 'map';
+
+interface ChemiscopeArgs {
+    dataset: Dataset;
+    height?: number;
+    width?: number | string;
+    selected_index?: number | undefined;
+    mode?: ChemiscopeMode;
+    settings?: Partial<Settings>;
+    no_info_panel?: boolean;
+}
+
+enum StreamlitValue {
+    SELECTION = 'selected_id',
+    SETTINGS = 'settings',
+}
+
+function getVisualizerClass(mode: ChemiscopeMode) {
+    switch (mode) {
+        case 'structure':
+            return StructureVisualizer;
+        case 'map':
+            return MapVisualizer;
+        default:
+            return DefaultVisualizer;
+    }
+}
+
+export class ChemiscopeComponent {
+    private state = {
+        visualizer: null as DefaultVisualizer | StructureVisualizer | MapVisualizer | null,
```

Comment:
I'm just seeing this, and it is fine to merge, but typically people use `undefined` for optional data, not `null`. There is a special syntax for it:

```suggestion
        visualizer?: DefaultVisualizer | StructureVisualizer | MapVisualizer,
```

Which is the same as 

```suggestion
        visualizer: DefaultVisualizer | StructureVisualizer | MapVisualizer | undefined,
```

---

I'm not sure what the `null as .... | null` does in this context as well, this is not how you set a default value

Reply:
agree, looks very nice and clean 😃